File: Day 2/Solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func LoadInput(fileName string) []map[string]int {

	file, err := os.Open(fileName)
	if err != nil {
		fmt.Errorf("Error reading in file %w", err)
	}

	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanLines)

	var lines []string

	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	var input []map[string]int

	// for each line in the input document, convert the command/value pair into a slice
	// of string/int pairs
	for _, line := range lines {

		valuesSlice := strings.Fields(line)
		command := valuesSlice[0]
		value, _ := strconv.Atoi(valuesSlice[1])

		tempMap := map[string]int{command: value}
		input = append(input, tempMap)
	}
	return input
}

func NavigationPart1(input *[]map[string]int) int {

	distance, depth := 0, 0

	for _, commandMap := range *input {

		for command, value := range commandMap {

			switch command {
			case "forward":
				distance += value
			case "up":
				depth -= value
			case "down":
				depth += value
			}
		}
	}

	return distance * depth
}

func NavigationPart2(input *[]map[string]int) int {

	distance, depth, aim := 0, 0, 0

	for _, commandMap := range *input {

		for command, value := range commandMap {

			switch command {
			case "forward":
				distance += value
				depth += value * aim
			case "up":
				aim -= value
			case "down":
				aim += value
			}
		}
	}

	return distance * depth
}

func main() {

	input := LoadInput("input.txt")

	fmt.Println(NavigationPart1(&input))
	fmt.Println(NavigationPart2(&input))

}

File: Day 2/Solution.py
from enum import Enum
from typing import Match

class CommandType(str, Enum):
    Forward = 'forward'
    Down    = 'down'
    Up      = 'up'

class Solution:

    def __init__(self, fileLocation):

        # read in commands/directions
        input = open(fileLocation)
        instructions = [line.split() for line in input.readlines()]
        self.commands = [(command, int(value)) for command, value in instructions]

    def NavigatePart1(self):

        distance = 0
        depth = 0

        # 1746616
        # 1741971043
        
        for command, value in self.commands:
            match command:
                case CommandType.Forward:
                    distance += value
                case CommandType.Up:
                    depth -= value
                case CommandType.Down:
                    depth += value
        
        return distance * depth

    def NavigatePart2(self):

        distance = 0
        depth = 0
        aim = 0
        
        for command, value in self.commands:
            match command:
                case CommandType.Forward:
                    distance += value
                    depth += value * aim
                case CommandType.Up:
                    aim -= value
                case CommandType.Down:
                    aim += value
        
        return distance * depth

solution = Solution("input.txt")
print(solution.NavigatePart1())
print(solution.NavigatePart2())